keps/sig-node/20190425-node-readiness-gates.md

```diff
@@ -0,0 +1,388 @@
+
+As Kubernetes becomes more and more customizable (or extensible) and gets deployed in mission critical systems,
+it is imperative to provide fail safe mechanisms that are built into the system.
+
+The remainder of this proposal introduces an approach for addressing the issue of node health extensibility using
+node readiness gates.
+
+### Goals
+
+* introduce a mechanism that allows for customizable/extensible node readiness checks.
+
+### Non-Goals
+
+* improving time-to-readiness of nodes
+
+## Proposal
+
+Any solution that is chosen for addressing the set of problems stated above should meet the following design requirements:
+
+* Be backwards compatible
+* Be extensible to accommodate variations in kubernetes environments - the set of system microservices can vary across k8s deployments.
+* Ideally, provide a single signal (not a policy) to cluster level components.
+* Ideally, avoid introducing new API primitives.
+* Support self-bootstrapping - scheduling system components (including the kubelet) even if the node is not logically ready.
+
+## Design Details
+
+### Determining Readiness via Pod Selectors
+
+In this design, the indication of node readiness can be expressed by the readiness/health of specific pods running on that node. This approach
+would leverage existing mechnanisms and well-known concepts to determine if a node is ready to receive pods.
+
+A readiness taint will be used to track the overall readiness of a node. The kubelet, node lifecycle controller, other external controllers will
+manage that taint.
+
+The Node object will be extended to include a list of readiness gates like so:
+
+```go
+type NodeSpec struct {
+  ...
+  // Defines the list of readiness gates for a node
+  ReadinessGates []NodeReadinessGate `json:"readinessGates,omitempty" protobuf:"bytes,7,opt,name=readinessGates"`
+}
+
+// NodeReadinessGate indicates a set of pods that must be ready on a given node before the
+// node itself is considered ready. Pods listed under NodeReadinessGates should be system critical pods
+// that should block other workloads from running on a node until it is ready.
+type NodeReadinessGate struct {
+   // The name of the readiness gate, mainly required as the patch merge key
+   Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+   // A label query for pods representing this readiness gate.
+   // Generally, labels here should point to a specific pod on a node.
+   Selector LabelSelector `json:"selector,omitempty" protobuf:"bytes,2,opt,name=selector"`
+   // The namespace query for the pods representing this readiness gate.
+   Namespace string `json:"namespace,omitempty" protobuf:"bytes,3,opt,name=namespace"`
+   // Indicates if this readiness gate should block the overall readiness of a node. Default: true
+   // +optional
+   BlocksReadiness bool `json:"blocksReadiness,omitempty" protobuf:"varint,4,opt,name=blocksReadiness"`
+}
+```
+
```

Comment:
Where are the ReadinessGates defined? If a new gate needs to be added what changes/configurations needs to be done and where?

Reply:
They are defined under `NodeSpec` (`spec` field in the node object). So you would add new readiness gates in `node.spec.readinessGates` which is a list of `NodeReadinessGate`s

Reply:
Sorry for being unclear in my question. I was wondering that when a new node is spawned how I should configure this node to have the ReadinessGates? I'm thinking for example an autoscaling array spawning new EC2 instances which acts as nodes. The fresh node needs to have these gates from the start to be useful. Will this be added as a new parameter into the kubelet command line options, or something similar?

Another option might be to create a new object into the api server which would act as an policy: the object would have a LabelSelector which would target the policy to matching nodes and then an array to define the ReadinessGates.

Reply:
Ah, good point. So my initial thoughts was that it would be done by the cloud provider and/or external controllers. The cloud provider case is easy because it can add node readiness gates during initialization. The external controller case is more tricky because the external controller may add the readiness gates after pods are already scheduled on a node. In that case a new flag or updates to component config might be required.

Reply:
> Another option might be to create a new object into the api server which would act as an policy: the object would have a LabelSelector which would target the policy to matching nodes and then an array to define the ReadinessGates.

I like this idea. I tried to avoid adding new APIs but the trade-off might be worth it? I'll let others comments on that one.

Reply:
It seems less coupling between kubelet and api server would be an advantage.

If details about the new object into the api server are outlined, we can have better idea between the two approaches.